code/modules/vehicles/mecha/mecha_construction_paths.dm

```diff
@@ -263,6 +300,89 @@
 		/obj/item/mecha_parts/part/ripley_right_leg
 	)
 
+/// Generic mech construction messages
+/datum/component/construction/mecha/custom_action(obj/item/I, mob/living/user, diff)
+	if(!..())
+		return FALSE
+
+	var/forward = (diff == FORWARD)
+	// Because of the way construction works, going backwards requires an offset to properly
+	// calculate which step we're on
+	var/curr_step = forward ? message_step : (message_step - 1)
+
+	// Skip over steps we're not doing!
+	if(!has_treads && curr_step == ADD_TREADS_STEP)
+		message_step += diff
+		curr_step += diff
+
+	if(!has_weapons_module)
+		if(curr_step == ADD_WEAPONS_CONTROLS_STEP || curr_step == SECURE_WEAPONS_CONTROLS_STEP)
+			message_step += diff * 2
+			curr_step += diff * 2
+
+	if(!has_bluespace_crystal)
+		if(curr_step == INSTALL_BLUESPACE_STEP || curr_step == CONNECT_BLUESPACE_STEP || curr_step == ENGAGE_BLUESPACE_STEP)
+			message_step += diff * 3
+			curr_step += diff * 3
```

Comment:
add a comment here about the order being important, since it is important

Reply:
I'm not sure exactly which part you mean. The order of the if blocks shouldn't matter here, since it's dependent on the step we're on.

Reply:
Lets say we're on ADD_TREADS_STEP, and we don't have treads.
Pretending for a moment there was another optional buildstep directly after treads, we enter that step.
If we checked for that step before we checked for treads, we would be in an illegal position

Reply:
What would happen if we skipped backwards into a step we didn't want? Wouldn't that have the same issue but the steps are out of order?

Reply:
yes actually, it seems to be kinda an issue with how your logic is setup.
maybe have a get_next_step() proc  with a while loop or something?

Reply:
The way that construction code works for messages pains me greatly, but that's probably a better idea.